io-engine/src/bdev/nexus/nexus_bdev_children.rs

```diff
@@ -330,36 +332,43 @@ impl<'n> Nexus<'n> {
         res
     }
 
-    /// offline a child device and reconfigure the IO channels
-    pub(crate) async fn offline_child(
+    /// Faults a child with the given reason.
+    pub async fn fault_child(
         mut self: Pin<&mut Self>,
         child_uri: &str,
+        reason: FaultReason,
     ) -> Result<NexusStatus, Error> {
-        info!("{:?}: offline child request: '{}'", self, child_uri);
+        info!(
+            "{:?}: fault ({}) child request for '{}'",
+            self, reason, child_uri
+        );
 
-        let paused = self.as_mut().pause_rebuild_jobs(child_uri).await;
+        self.check_nexus_operation(NexusOperation::ReplicaFault)?;
 
-        let res = match self.as_mut().child_mut(child_uri) {
-            Ok(child) => {
-                child.offline().await;
-                self.reconfigure(DrEvent::ChildOffline).await;
-                Ok(self.status())
-            }
-            Err(e) => Err(e),
-        };
+        let dev_name = self.get_child_device_name(child_uri)?;
 
-        paused.resume().await;
+        // TODO: check to prevent faulting if a rebuild is running (?).
+        self.as_mut().cancel_rebuild_jobs(child_uri).await;
 
-        res
+        // TODO: check to prevent the last child from degrading (?).
+        self.as_mut().retire_child_device(&dev_name, reason, false);
+
+        Ok(self.status())
     }
 
-    /// Faults a child device and reconfigures the IO channels.
-    pub async fn fault_child(
+    /// TODO: Legacy method to fault child by client.
+    pub async fn fault_child_legacy(
         mut self: Pin<&mut Self>,
         child_uri: &str,
-        reason: Reason,
     ) -> Result<(), Error> {
-        info!("{:?}: fault child request for '{}'", self, child_uri);
+        self.check_nexus_operation(NexusOperation::ReplicaFault)?;
+
+        let reason = FaultReason::OfflinePermanent;
```

Comment:
Is control plane expected to differentiate between Offline and OfflinePermanent before removing the child after the fault? i.e. legacy behaviour of removing the child immediately upon seeing FaultReason::OfflinePermanent

Reply:
I preserved the current logic, just renamed internal states. Probably not the best names, but I'm going to refactor that anyway.
They are mapped into different API child states:
- Offline  => CHILD_STATE_DEGRADED + CHILD_STATE_REASON_BY_CLIENT
- OfflinePermanent => CHILD_STATE_FAULTED + CHILD_STATE_REASON_BY_CLIENT